src/charm.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm Jenkins."""
+
+import logging
+from pathlib import Path
+from time import sleep
+from typing import TYPE_CHECKING, Any
+
+import requests
+from jenkinsapi.jenkins import Jenkins
+from ops.charm import CharmBase, PebbleReadyEvent
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus
+from ops.pebble import Layer
+
+from types_ import Credentials
+
+if TYPE_CHECKING:
+    from ops.pebble import LayerDict  # pragma: no cover
+
+logger = logging.getLogger(__name__)
+
+JENKINS_WEB_URL = "http://localhost:8080"
+JENKINS_HOME = Path("/var/lib/jenkins")
+# Path to initial admin password file
+INITIAL_PASSWORD = JENKINS_HOME / Path("secrets/initialAdminPassword")
+# Path to last executed jenkins version file, required to override wizard installation
+LAST_EXEC = JENKINS_HOME / Path("jenkins.install.InstallUtil.lastExecVersion")
+# Path to jenkins version file, required to override wizard installation
+UPDATE_VERSION = JENKINS_HOME / Path("jenkins.install.UpgradeWizard.state")
+
+
+class JenkinsK8SOperatorCharm(CharmBase):
```

Comment:
Should we use the new charm architecture guidance for this charm?

Reply:
Updated, will leave the conversations unresolved until it gets approval :)